README.md
# ACI Modular L3Out

[![published](https://static.production.devnetcloud.com/codeexchange/assets/images/devnet-published.svg)](https://developer.cisco.com/codeexchange/github/repo/netcloudag/AciL3outModular)

Python Module for configuring ACI L3Outs on a Modular bases.
Currently Supported is Static- and BGP-Routing (inclusive Multicast). 
The Module Provide an abstraction Layer in Form of Python Objects and gives back 
the JSON which could be posted to the APIC.

## Getting Started

Install Module via pip install or git clone. Please see Section Installing below...

```
from L3Out import ModularL3Out
L3Out2Post = ModularL3Out.L3Out("L3Out-Name-xyz", "Tenant-Name-xyz")
etc...
print L3Out2Post.tostring()
```

### Prerequisites

What things you need to install the software and how to install them

```
Python 3.7.X
```

### Installing

```
pip install --upgrade git+https://github.com/netcloudag/AciL3outModular.git
```

Or

```
git clone https://github.com/netcloudag/AciL3outModular.git
cd AciL3outModular
pip install -r requirements.txt
```

## Running the tests

Simple UnitTest is under test Folder

## Contributing

Please read [CONTRIBUTING.md](https://github.com/netcloudag/AciL3outModular/blob/master/CONTRIBUTING.md) for details on our code 
of conduct, and the process for submitting pull requests to this project.

## Versioning

For the versions available, see the [tags on this repository](https://github.com/netcloudag/AciL3outModular/tags). 

## Authors

* ** Andreas Graber ** - *Initial work*

## License

Please - see the [LICENSE.md](https://github.com/netcloudag/AciL3outModular/blob/master/LICENSE.md) file for details. 


## Example with Sandbox https://sandboxapicdc.cisco.com

This example is done with the script examples/script_eBGP.py
APIC Access-Credential must be inserted in examples/settings.conf like:

```
[APIC]
ip = sandboxapicdc.cisco.com
user = admin
password = PwdOfSandboxApic
```

The APIC Tenant must be existing!

The Script can be tested on all Cisco ACI APICs including the Simulator. The Supported APIC Versions are 2, 3 and 4.

In the beginning the Sandbox APIC is configured without the L3Outs:

![Fabric Tenant](/docs/images/clean_fabric.jpg)

Then we run the script:

![Script Output](/docs/images/script_run.jpg)

The Result in the Sandbox APIC is:

![L3Out posted](/docs/images/after_post.jpg)






CHANGELOG.md
# Changelog
> * v1.0.0, 21.11.2019 -- Initial release - AGR
> * v1.1.0, 15.04.2020 -- Changed to Python 3 - AGR


test/TestModularL3Out.py
# <editor-fold desc="Imports">
import unittest
from L3Out.ModularL3Out import L3Out
# </editor-fold>


class TestModularL3Out(unittest.TestCase):

    def test_simple(self):
        l3_out_2_post = L3Out("L3Out-Name-xyz", "Tenant-Name-xyz")
        l3_out_2_post.setEnableBgp()
        l3_out_2_post.setExternalEpg('External')
        l3_out_2_post.externalEpg().setL3ExtSubnet('0.0.0.0/0', False, False)
        l3_out_2_post.setNodeProfile('Node')
        l3_out_2_post.nodeProfile().setNode('topology/pod-1/node-111','10.1.1.111')
        l3_out_2_post.nodeProfile().setInt('INT')
        l3_out_2_post.nodeProfile().Int().setIntNode('sub-interface', "vlan-123", '10.1.2.1/30', 'topology/pod-1/paths-111/pathep-[eth1/50]')
        l3_out_2_post.nodeProfile().Int().intNode().setBgpPeer('10.1.2.2', 'Secure-PWD')
        l3_out_2_post.nodeProfile().Int().intNode().bgpPeer().setBgpAS('65412')

        test_dict = {'l3extOut': {'attributes': {'dn': 'uni/tn-Tenant-Name-xyz/out-L3Out-Name-xyz', 'name': 'L3Out-Name-xyz', 'rn': 'out-L3Out-Name-xyz', 'status': 'created,modified'}, 'children': [{'bgpExtP': {'attributes': {'rn': 'bgpExtP', 'status': 'created,modified'}}}, {'l3extInstP': {'attributes': {'name': 'External', 'rn': 'instP-External', 'status': 'created,modified'}, 'children': [{'l3extSubnet': {'attributes': {'aggregate': '', 'ip': '0.0.0.0/0', 'rn': 'extsubnet-[0.0.0.0/0]', 'status': 'created,modified'}}}]}}, {'l3extLNodeP': {'attributes': {'name': 'Node', 'rn': 'lnodep-Node', 'status': 'created,modified'}, 'children': [{'l3extRsNodeL3OutAtt': {'attributes': {'rn': 'rsnodeL3OutAtt-[topology/pod-1/node-111]', 'rtrIdLoopBack': 'no', 'tDn': 'topology/pod-1/node-111', 'rtrId': '10.1.1.111', 'status': 'created,modified'}, 'children': [{'l3extInfraNodeP': {'attributes': {'fabricExtCtrlPeering': 'false', 'status': 'created,modified'}, 'children': []}}]}}, {'l3extLIfP': {'attributes': {'rn': 'lifp-INT', 'name': 'INT', 'status': 'created,modified'}, 'children': [{'l3extRsPathL3OutAtt': {'attributes': {'mac': '00:22:BD:F8:19:FF', 'ifInstT': 'sub-interface', 'encap': 'vlan-123', 'addr': '10.1.2.1/30', 'tDn': 'topology/pod-1/paths-111/pathep-[eth1/50]', 'rn': 'rspathL3OutAtt-[topology/pod-1/paths-111/pathep-[eth1/50]]', 'status': 'created,modified'}, 'children': [{'bgpPeerP': {'attributes': {'rn': 'peerP-[10.1.2.2]', 'addr': '10.1.2.2', 'password': 'Secure-PWD', 'status': 'created,modified'}, 'children': [{'bgpAsP': {'attributes': {'rn': 'as', 'asn': '65412', 'status': 'created,modified'}, 'children': []}}]}}]}}]}}]}}]}}

        self.assertDictEqual(l3_out_2_post.tostring(), test_dict)


if __name__ == '__main__':
    unittest.main()


examples/script_eBGP.py
# ==============================================================================
from L3Out import ModularL3Out
from configparser import ConfigParser
import os
import sys
import requests
import json
import openpyxl
import warnings
import urllib3
warnings.filterwarnings("ignore", category=DeprecationWarning)
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
# ==============================================================================

# ==============================================================================
# changeDir
# ==============================================================================
if os.path.dirname(sys.argv[0]) != "":
    os.chdir(os.path.dirname(sys.argv[0]))

# ==============================================================================
# init ACI & login
# ==============================================================================
config = ConfigParser()
config.read("settings.conf")
apicIp = config.get("APIC", "ip")
apicUrl = 'https://' + apicIp + '/api/'
apicUser = config.get("APIC", "user")
apicPw = config.get("APIC", "password")

# create reqeuests session
session = requests.Session()

# create credentials structure
userPass = json.dumps({'aaaUser': {'attributes': {'name': apicUser, 'pwd': apicPw}}})

# login to API
response = session.post(apicUrl + 'aaaLogin.json', data=userPass, verify=False, timeout=10)

token = None

# Error Handling
if response.status_code == 401:
    raise Exception('Unauthorized')
    # Raise a exception for all other 4xx and 5xx status_codes
elif response.raise_for_status():
    raise Exception('Error occured: ' + response.status_code)
else:
    token = response.json()['imdata'][0]['aaaLogin']['attributes']['token']


# ==============================================================================
# postJson
# ==============================================================================
def postJson(jsonData, url='mo.json'):
    response = session.post(apicUrl + url, verify=False, data=json.dumps(jsonData, sort_keys=True))

    if response.raise_for_status:
        if response.status_code == 200:
            return response.status_code
        elif response.status_code == 400:
            return '400: ' + response.json()['imdata'][0]['error']['attributes']['text']
        else:
            return response.status_code


# ==============================================================================
# check excel-content
# ==============================================================================
def check_set(content):
    if content != "None" and content != "" and content is not None:
        return True
    else:
        return False


def get_excel_data():
    xlsxFo = "L3OUT_eBGP.xlsx"
    wbObj = openpyxl.load_workbook(xlsxFo, data_only=True)
    sheet = wbObj.get_sheet_by_name("L3OUT")

    L3Out = []
    Keys = []

    for row in range(2, sheet.max_row + 1):
        if row == 2:
            for col in range(1,sheet.max_column+1):
                value = sheet.cell(row=row, column=col).value
                if value != None: Keys.append(value)
            # print(Keys)
        else:
            if check_set(str(sheet["A" + str(row)].value)):
                line = {}
                for x in range(len(Keys)):
                    line[Keys[x]]=str(sheet.cell(row=row, column=x+1).value)
                L3Out.append(line)
    return L3Out


def get_remote_bgp_ip(ip):
    BGP_IP = ip.partition(("/")[0])[0].split(".")
    BGP_IP[3] = int(BGP_IP[3]) - 1
    return '.'.join(str(e) for e in BGP_IP)


if __name__ == '__main__':

    Tenant = "ModularL3Out"
    Contract = "CT-PERMIT-ALL"

    Path1 = "topology/pod-1/node-111"
    Paths1 = "topology/pod-1/paths-111/pathep-[eth1/50]"

    Path2 = "topology/pod-1/node-112"
    Paths2 = "topology/pod-1/paths-112/pathep-[eth1/50]"

    Path3 = "topology/pod-1/node-211"
    Paths3 = "topology/pod-1/paths-211/pathep-[eth1/50]"

    Path4 = "topology/pod-1/node-212"
    Paths4 = "topology/pod-1/paths-212/pathep-[eth1/50]"

    Lo1 = "10.10.20.128"
    Lo2 = "10.10.20.129"
    Lo3 = "10.10.20.142"
    Lo4 = "10.10.20.143"

    BGP_PWD = "SecureTest"
    AST1 = "65164"
    LocalAS = "65162"

    L3Out_Data = get_excel_data()

    for item in L3Out_Data:

        L3Out2Post = ModularL3Out.L3Out(item["NAME"], Tenant)
        L3Out2Post.setl3domain(item["Dom"])
        L3Out2Post.setVrf(item["VRF"])
        L3Out2Post.setExternalEpg(item["Ext-EPG"].split(":")[0])
        L3Out2Post.externalEpg().setConsumeContract(Contract)
        L3Out2Post.externalEpg().setProvideContract(Contract)
        L3Out2Post.externalEpg().setL3ExtSubnet()
        L3Out2Post.setEnableBgp()

        L3Out2Post.setNodeProfile(item["NP"])
        L3Out2Post.nodeProfile().setNode(Path1, Lo1)
        L3Out2Post.nodeProfile().setNode(Path2, Lo2)
        L3Out2Post.nodeProfile().setNode(Path3, Lo3)
        L3Out2Post.nodeProfile().setNode(Path4, Lo4)

        L3Out2Post.nodeProfile().setInt(item["INT"])
        L3Out2Post.nodeProfile().Int().setIntNode(item["TYP"], "vlan-" + item["VLAN1"], item["P-IP1"] + "/30", Paths1)

        RemoteIP = get_remote_bgp_ip(item["P-IP1"])

        L3Out2Post.nodeProfile().Int().intNode().setBgpPeer(RemoteIP, BGP_PWD)
        L3Out2Post.nodeProfile().Int().intNode().bgpPeer().setBgpAS(AST1)
        L3Out2Post.nodeProfile().Int().intNode().bgpPeer().setBgpLocalAS(LocalAS)

        L3Out2Post.nodeProfile().Int().setIntNode(item["TYP"], "vlan-" + item["VLAN2"], item["P-IP2"] + "/30", Paths2)

        RemoteIP = get_remote_bgp_ip(item["P-IP2"])

        L3Out2Post.nodeProfile().Int().intNode().setBgpPeer(RemoteIP, BGP_PWD)
        L3Out2Post.nodeProfile().Int().intNode().bgpPeer().setBgpAS(AST1)

        L3Out2Post.nodeProfile().Int().setIntNode(item["TYP"], "vlan-" + item["VLAN3"], item["P-IP3"] + "/30", Paths3)

        RemoteIP = get_remote_bgp_ip(item["P-IP3"])

        L3Out2Post.nodeProfile().Int().intNode().setBgpPeer(RemoteIP, BGP_PWD)
        L3Out2Post.nodeProfile().Int().intNode().bgpPeer().setBgpAS(AST1)

        L3Out2Post.nodeProfile().Int().setIntNode(item["TYP"], "vlan-" + item["VLAN4"], item["P-IP4"] + "/30", Paths4)

        RemoteIP = get_remote_bgp_ip(item["P-IP4"])

        L3Out2Post.nodeProfile().Int().intNode().setBgpPeer(RemoteIP, BGP_PWD)
        L3Out2Post.nodeProfile().Int().intNode().bgpPeer().setBgpAS(AST1)
        L3Out2Post.nodeProfile().Int().intNode().bgpPeer().setBgpLocalAS(LocalAS)

        print('Connecting To: ' + apicUrl + ' and posting: \n' + str(L3Out2Post.tostring()))
        response = postJson(L3Out2Post.tostring())
        print('Response-Code is: ' + str(response))


